include/beast/core/impl/flat_streambuf.ipp

```diff
@@ -0,0 +1,314 @@
+//
+// Copyright (c) 2013-2016 Vinnie Falco (vinnie dot falco at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef BEAST_IMPL_FLAT_STREAMBUF_HPP
+#define BEAST_IMPL_FLAT_STREAMBUF_HPP
+
+#include <boost/assert.hpp>
+#include <stdexcept>
+
+namespace beast {
+
+/*  Memory is laid out thusly:
+
+    p_ ..|.. in_ ..|.. out_ ..|.. last_ ..|.. end_
+*/
+
+template<class Allocator>
+basic_flat_streambuf<Allocator>::
+~basic_flat_streambuf()
+{
+    if(p_)
+        alloc_traits::deallocate(
+            this->member(), p_, dist(p_, end_));
+}
+
+template<class Allocator>
+basic_flat_streambuf<Allocator>::
+basic_flat_streambuf(basic_flat_streambuf&& other)
+    : detail::empty_base_optimization<
+        allocator_type>(std::move(other.member()))
+{
+    p_ = other.p_;
+    in_ = other.in_;
+    out_ = other.out_;
+    last_ = out_;
+    end_ = other.end_;
+    max_ = other.max_;
+    other.p_ = nullptr;
+    other.in_ = nullptr;
+    other.out_ = nullptr;
+    other.last_ = nullptr;
+    other.end_ = nullptr;
+}
+
+template<class Allocator>
+basic_flat_streambuf<Allocator>::
+basic_flat_streambuf(basic_flat_streambuf&& other,
+        Allocator const& alloc)
+    : detail::empty_base_optimization<
+        allocator_type>(alloc)
+{
+    if(this->member() != other.member())
+    {
+        auto const n = other.size();
+        p_ = alloc_traits::allocate(
+            this->member(), n);
+        in_ = p_;
+        out_ = p_ + n;
+        last_ = out_;
+        end_ = out_;
+        max_ = other.max_;
+        std::memcpy(in_, other.in_, n);
+        return;
+    }
+    p_ = other.p_;
+    in_ = other.in_;
+    out_ = other.out_;
+    last_ = out_;
+    end_ = other.end_;
+    max_ = other.max_;
+    other.p_ = nullptr;
+    other.in_ = nullptr;
+    other.out_ = nullptr;
+    other.last_ = nullptr;
+    other.end_ = nullptr;
+}
+
+template<class Allocator>
+basic_flat_streambuf<Allocator>::
+basic_flat_streambuf(
+        basic_flat_streambuf const& other)
+    : detail::empty_base_optimization<allocator_type>(
+        alloc_traits::select_on_container_copy_construction(
+            other.member()))
+{
+    auto const n = other.size();
+    p_ = alloc_traits::allocate(
+        this->member(), n);
+    in_ = p_;
+    out_ = p_ + n;
+    last_ = out_;
+    end_ = out_;
+    max_ = other.max_;
+    std::memcpy(in_, other.in_, n);
```

Comment:
Are there any circumstances when `other.in_` can be `nullptr`? If so, this will have UB even if `n == 0` (see: http://en.cppreference.com/w/cpp/string/byte/memcpy)